### ISSUE TITLE
1.3.3版本开始，SINGLE_PROCESS 和 CRYPT下，encode一个数据后，decode的这个key的数据可能是错的

### ISSUE THREAD

**Issue by XXUUNN** [Tencent/MMKV#1351]
### The version of MMKV
 **v1.3.3 &  version-latest**

### The platform of MMKV
 Android

### What's the issue?
1.3.3版本开始，SINGLE_PROCESS 和 CRYPT下，encode一个数据后，decode这个key的数据很可能出现错误，decode失败。
有错误输出 <MMKV.cpp:744::getBytes> InvalidProtocolBuffer truncatedMessage
- 1.3.3之前版本(1.3.2、1.2.15)简单测试没出现。
- 不加密或者多进程下简单测试没出现。
```
    boolean getOrSet = true;
    String s;
    MMKV testMMKV;
    private void test12logWriteAndUpload(boolean b) {
        if (true) {
            String key = "nn1";
            String name = "mmyu111";
            if (getOrSet) {
                if (testMMKV == null) {
                    String absolutePath = getExternalFilesDir("00000").getAbsolutePath();
                    testMMKV = MMKV.mmkvWithID(name, MMKV.SINGLE_PROCESS_MODE, "f", absolutePath);
                }
                try {
                    if (true) {
                        throw new Exception("x");
                    }
                    s = testMMKV.decodeString(key, "");
                } catch (Exception e) {
                    byte[] bytes = testMMKV.decodeBytes(key, "".getBytes(Charset.forName("UTF-8")));
                    if (bytes != null) {
                        try {
                            String s1 = new String(bytes, Charset.forName("UTF-8"));
                            s = s1;
                        } catch (Exception ex) {
                            Log.e(TAG, "test12logWriteAndUpload: bytes " +  Arrays.toString(bytes));
                            return;
                        }
                    }
                }
                Log.e(TAG, "test12logWriteAndUpload: <<get[" + s);
            } else {
                if (s.length() == 0) {
                    s = "11";
                } else {
                    char c = s.charAt(0);
                    int i = c - '0' + 0;
                    if (++i > 9) {
                        i = 0;
                    }
                    s = "" + i + i + I + i + i + I  + i + i + I  + i + i + I   + i + i + i + s;
                }
                Log.e(TAG, "test12logWriteAndUpload: >>set[" + s);
                testMMKV.encode(key, s);
            }
            getOrSet = !getOrSet;
            return;
        }
    }
```
    
### logcat output
``` text
2024-07-11 14:36:31.586  2450-2450  I  <MMKV_Android.cpp:158::mmkvWithID> prepare to load mmyu111 (id 308a9dc23d64e466418b84a8b489329a) from rootPath /storage/emulated/0/Android/data/com.tencent.mmkvdemo/files/00000
2024-07-11 14:36:31.587  2450-2450  I  <MemoryFile.cpp:97::open> open fd[0x90], /storage/emulated/0/Android/data/com.tencent.mmkvdemo/files/00000/308a9dc23d64e466418b84a8b489329a
2024-07-11 14:36:31.587  2450-2450  I  <MemoryFile.cpp:202::mmap> mmap to address [0x7982d65000], oldPtr [0x0], [/storage/emulated/0/Android/data/com.tencent.mmkvdemo/files/00000/308a9dc23d64e466418b84a8b489329a]
2024-07-11 14:36:31.587  2450-2450  I  <MemoryFile.cpp:97::open> open fd[0x9d], /storage/emulated/0/Android/data/com.tencent.mmkvdemo/files/00000/308a9dc23d64e466418b84a8b489329a.crc
2024-07-11 14:36:31.587  2450-2450  I  <MemoryFile.cpp:202::mmap> mmap to address [0x7982d64000], oldPtr [0x0], [/storage/emulated/0/Android/data/com.tencent.mmkvdemo/files/00000/308a9dc23d64e466418b84a8b489329a.crc]
2024-07-11 14:36:31.587  2450-2450  I  <MMKV_IO.cpp:210::loadMetaInfoAndCheck> meta file [308a9dc23d64e466418b84a8b489329a] has flag [0]
2024-07-11 14:36:31.587  2450-2450  I  <MMKV_IO.cpp:80::loadFromFile> loading [308a9dc23d64e466418b84a8b489329a] with 269 actual size, file size 4096, InterProcess 0, meta info version:4
2024-07-11 14:36:31.587  2450-2450  I  <MMKV_IO.cpp:85::loadFromFile> loading [308a9dc23d64e466418b84a8b489329a] with crc 4294277947 sequence 1 version 4
2024-07-11 14:36:31.587  2450-2450  I  <MMKV_IO.cpp:129::loadFromFile> loaded [308a9dc23d64e466418b84a8b489329a] with 1 key-values
2024-07-11 14:36:31.587  2450-2450  E  test12logWriteAndUpload: <<get[88888888888888877777777777777766666666666666655555555555555544444444444444433333333333333322222222222222211111111111111100000000000000099999999999999988888888888888877777777777777766666666666666655555555555555544444444444444433333333333333322222222222222211
2024-07-11 14:36:31.761  2450-2450  E  test12logWriteAndUpload: >>set[99999999999999988888888888888877777777777777766666666666666655555555555555544444444444444433333333333333322222222222222211111111111111100000000000000099999999999999988888888888888877777777777777766666666666666655555555555555544444444444444433333333333333322222222222222211
2024-07-11 14:36:31.921  2450-7127  I  Compiler allocated 4991KB to compile void android.view.ViewRootImpl.performTraversals()
2024-07-11 14:36:31.936  2450-2450  E  <MMKV.cpp:744::getBytes> InvalidProtocolBuffer truncatedMessage
2024-07-11 14:36:31.936  2450-2450  E  test12logWriteAndUpload: <<get[
2024-07-11 14:36:32.107  2450-2450  E  test12logWriteAndUpload: >>set[11
2024-07-11 14:36:32.276  2450-2450  E  test12logWriteAndUpload: <<get[11
2024-07-11 14:36:32.428  2450-2450  E  test12logWriteAndUpload: >>set[22222222222222211
2024-07-11 14:36:32.592  2450-2450  E  test12logWriteAndUpload: <<get[22222222222222211
2024-07-11 14:36:32.741  2450-2450  E  test12logWriteAndUpload: >>set[33333333333333322222222222222211
```

**Comment by lingol** [comment]
I can not reproduce this problem. Please provide the whole demo project. For example your `I` is not defined.

**Comment by lingol** [comment]
Wait I have successfully reproduced this error after retrying many times. 
Will look into it next week.

**Comment by lingol** [comment]
It's fixed in the dev branch. This bug was introduced in v1.3.3 in this optimization:
> Optimizing write speed when there's only one key inside MMKV.

**Comment by lingol** [comment]
It's released in [v1.3.9](https://github.com/Tencent/MMKV/releases/tag/v1.3.9).
